Yul: custom source locations (`@src`)

christianparpart opened the pull request:
(the branch name is misleading)

This implements the parser part of #11086.

I've added some tests to make sure nesting does produce the expected source locations, I am however not too sure if I should test now every type of AST node for Yul. Should I?

### checklist

- [x] eliminate `_fromDoc` (make it part of `Parser` instead
- [x] refert the `{}`'s in AST.h
- [x] if an `@src` was found, use that not just for the next AST node.
- [x] write tests for newly added parser errors

christianparpart committed:
Adds missing include header (for DebugData).

christianparpart committed:
Yul: Adds parsing @src comment in AsmParser to customize the AST's sourcer locations.

chriseth commented on libyul/AsmParser.h:
```diff
@@ -48,30 +50,56 @@ class Parser: public langutil::ParserBase
 		Dialect const& _dialect,
-		std::optional<langutil::SourceLocation> _locationOverride = {}
+		std::optional<langutil::SourceLocation> _locationOverride = {},
+		std::function<std::shared_ptr<langutil::CharStream>(unsigned)> _charStreamForSourceIndex = {}
```
Please document.

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -81,6 +81,7 @@
 #include <utility>
 #include <map>
 #include <range/v3/view/concat.hpp>
+#include <range/v3/view/transform.hpp>
```
Needed?

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -927,6 +928,12 @@ map<string, unsigned> CompilerStack::sourceIndices() const
 
+langutil::CharStream const* CompilerStack::charStreamForSourceIndex(unsigned _sourceIndex) const
+{
+	solAssert(_sourceIndex < m_sources.size(), "Cannot get CharStream of higher index than sources available.");
```
This relies on the way `sourceIndices()` works without assuring it. It also ignores the `yulUtilityFileName`. Can you use a common data source for both functions? It does not have to be fast, the number of source files is less than 100 I would say.

chriseth commented on libyul/AST.h:
```diff
@@ -45,7 +45,7 @@ struct DebugData
 };
 
-struct TypedName { std::shared_ptr<DebugData const> debugData; YulString name; Type type; };
+struct TypedName { std::shared_ptr<DebugData const> debugData; YulString name{}; Type type{}; };
```
I think making everything except the debug data optional is very dangerous.

christianparpart replied:
What do you suggest. Also adding `{}` to `debugData` or removing it again? (the latter would cause quite a few inconveniences in Parser.cpp. no blocker though)

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,33 +47,63 @@ class Parser: public langutil::ParserBase
-		T r;
-		r.debugData = DebugData::create(currentLocation());
-		return r;
+		auto const fromDocumentation = updateLocation();
+		return std::tuple<T, bool>{
+			T{DebugData::create(currentLocation()), std::forward<Args>(_args)...},
+			fromDocumentation
+		};
```
Why can't you use `T r; T.debugData = ...` as it was before?

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,33 +47,63 @@ class Parser: public langutil::ParserBase
 	std::unique_ptr<Block> parse(std::shared_ptr<langutil::Scanner> const& _scanner, bool _reuseScanner);
 
 protected:
+	std::tuple<langutil::SourceLocation, bool> currentOverridableLocation() const
```
Please document the difference to `currentLocation`. Maybe it is easier to understand if you rename `currentLocation` to `currentScannerLocation`? (Is that what it is?)

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,16 +64,77 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+shared_ptr<DebugData const> Parser::updateLocationEndFrom(
+	shared_ptr<DebugData const> const& _debugData,
+	langutil::SourceLocation const& _location,
+	bool _fromDoc
```
Shouldn't `_fromDoc` be a switch that is activated once in the parser when the first tag it sees is `@use-src` and disabled otherwise?

christianparpart replied:
I think it is possible to have source locations mixed. I think when you parse yul code some parts can be solidity transformed yul code, and some auto-generated or injected, isn't it?

christianparpart replied:
IIRC we agreed to not allow mixed modes for now. the other is implemented. hiding & marking as resolved.

chriseth commented:
Please check (using a test) that the source location change is correctly applied in the following example:
```
{
  {
  /// @src a:b;c
  }
  let x := 1
}
```

If it is not, let us talk about it, it might be difficult.

chriseth commented:
I think the above can be realized in the following way:

1) the parser has to be created in one of two modes: a) taking source locations from the scanner b) taking source locations from comments - in the second case, the index -> source mapping is passed to the constructor. The `@use-src` tag switches the parser into this mode, but it has to be the very first thing in the source. Because of this, it could be outside the code (in the object). This, the object parser has to deal with it.

2) if in "take from comments" mode, the `advance()` function has to be overridden (and we have to make sure it is always used), such that it checks if after the advancement, there is a "current comment literal" containing a `@src` annotation. In this case, it is stored as the current source location (we could re-use "m_locationOverride" but would probable rename it). We have to make sure that this also works at the beginning of parsing, before the first call to `advance()`.

No further change to the parser is required.

chriseth commented:
Actually there is one tiny change: `updateLocationEndFrom` has to be omitted if we have a location from comments. This should have been the case for `m_locationOverride` already and is kind of a bug, but it is irrelevant if the "override location" is always the same.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -131,24 +203,27 @@ Statement Parser::parseStatement()
+		auto stmt = createWithLocation<Break>();
 		checkBreakContinuePosition("break");
 		advance();
+		stmt.debugData = updateLocationEndFrom(stmt.debugData, currentLocation());
```
Why are these needed?

chriseth commented:
Can you extract the changes to ParserBase and the Solidity Parser to its own pull request?

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -131,11 +135,10 @@ bool Parser::updateLocation()
 )
 {
-	if (_fromDoc)
+	if (m_locationOverride)
```
I would prefer the `if` to be at the call site.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,6 +64,65 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	// TODO: [ ] updateLocationEndFrom(): pull top if-condition out of it
+
+	static std::regex const lineRE = std::regex(
+		R"~~~(.*@src\s+(\d+):(-1|\d+):(-1|\d+).*)~~~",
```
I think `.*` at the start is not useful because it would also parse `x@src`

Also why `.*` at the end?

christianparpart replied:
changed vom `.*` to `\s*`.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,6 +64,65 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	try
+	{
+		solAssert(cm.size() == 4, "");
+		auto const sourceIndex = static_cast<unsigned>(stoul(cm[1].str()));
```
does the static_cast truncate?

christianparpart replied:
stoul uses strtoul which sets errno ERANGE which will be translated to `std::out_of_range{}` It will return an `unsigned long` and we expect just an `unsigned int`, so it will be truncated if within range of ulong. I think I'll guard against that case.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,6 +64,65 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+
+		m_locationOverride = SourceLocation{start, end, charStream};
+	}
+	catch (std::out_of_range) {}
```
Just ignore?

chriseth replied:
I would propose a parse error.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -131,21 +177,21 @@ Statement Parser::parseStatement()
 	case Token::Break:
 	{
-		Statement stmt{createWithLocation<Break>()};
+		Break stmt = createWithLocation<Break>();
```
Why the change?

christianparpart replied:
was left over from some old code that required it. reverted.

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,13 +47,19 @@ class Parser: public langutil::ParserBase
 
+	/// Constructs a Yul parser.
+	///
+	/// @p _charStreamForSourceIndex a callback for retrieving the CharStream related to a source
```
Please update

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,13 +47,19 @@ class Parser: public langutil::ParserBase
 		Dialect const& _dialect,
-		std::optional<langutil::SourceLocation> _locationOverride = {}
+		std::optional<langutil::SourceLocation> _locationOverride = {},
+		std::optional<std::map<unsigned, std::shared_ptr<langutil::CharStream>>> _sourceMap = std::nullopt
```
I think `sourceMap` is can be easily confused with the bytecode-to-source-mapping

Also please document that this changes the way the parser works with regards to source locations.

chriseth commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -927,6 +926,17 @@ map<string, unsigned> CompilerStack::sourceIndices() const
+	unsigned index = 0;
+	for (auto const& s: m_sources)
+		result[index++] = s.second.scanner->charStream();
+	// NB: CompilerContext::yulUtilityFileName() does not have a source,
```
After some more steps into completing this feature, we won't have the source of the others either. So maybe just add it with an empty shared pointer?

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
 	while (currentToken() != Token::RBrace)
 		block.statements.emplace_back(parseStatement());
-	block.debugData = updateLocationEndFrom(block.debugData, currentLocation());
+	if (!m_locationOverride)
```
I'm not sure this is the right switch. What happens if we have
```
{
/// @src ...
let x := 1
}
```

The value of `!m_locationOverride` changes while we are parsing the block.

chriseth replied:
It might be that we need to distinguish three settings:
 - always take source location from scanner
 - always take source location from "location override" (and do not update it during parsing)
 - use empty source location at start (also not taking it from the scanner) and change it for every `@src`.

christianparpart replied:
we have an enum class for that now.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+		return;
+
+	static std::regex const lineRE = std::regex(
+		R"~~~(\s*@src\s+(\d+):(-1|\d+):(-1|\d+)\s*)~~~",
```
This still succeeds for `x@src 1:1:1`

christianparpart replied:
ah, due to the missing leading `^` which I removed earlier.

chriseth replied:
`@a @src 1:1:1` should also be valid.

christianparpart replied:
so `.*\s@src......`? that should consume anything ahead.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+			// Invalid source index is currently silently ignored.
+			// The SourceLocation from the Scanner will be used instead.
+			m_errorReporter.syntaxError(2674_error, m_scanner->currentLocation(), "Invalid source mapping. Source index not defined via @use-src.");
+			m_locationOverride.reset();
```
If you reset it here, in the current logic, this will lead to ends being assigned from the scanner...

christianparpart replied:
i think that's fixed with the new enum class.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+
+		m_locationOverride = SourceLocation{start, end, charStream};
+	}
+	catch (std::out_of_range const&)
```
Is it useful to make the try-range that large?

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+		{
+			// Invalid source index is currently silently ignored.
+			// The SourceLocation from the Scanner will be used instead.
+			m_errorReporter.syntaxError(2674_error, m_scanner->currentLocation(), "Invalid source mapping. Source index not defined via @use-src.");
```
I think the location already points to after the comment. Is it possible to get the location of the comment?

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +81,70 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+		solAssert(sourceIndex <= std::numeric_limits<unsigned long>::max(), "");
+		auto const start = stoi(cm[2].str());
+		auto const end = stoi(cm[3].str());
+		solAssert(start <= end, "");
```
I don't think we can assert that.

christianparpart replied:
so add an error report for that, too. okay.

chriseth commented on libyul/AsmParser.h:
```diff
@@ -47,20 +47,37 @@ class Parser: public langutil::ParserBase
+		ParserBase(_errorReporter),
+		m_dialect(_dialect),
+		m_locationOverride(std::move(_locationOverride)),
+		m_useSourceLocationFrom{m_locationOverride ?
```
newline after `{`

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,74 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	);
+
+	auto const commentLocation = m_scanner->currentCommentLocation();
+	string_view text = m_scanner->currentCommentLiteral();
```
Why not `string const&`?

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,74 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	auto const start = toInt(cm[2].str());
+	auto const end = toInt(cm[3].str());
+
+	if (!sourceIndex || !start || !end)
```
Maybe better to construct the SourcLocation object and then call `isValid` on it?

chriseth replied:
Ah sorry, these are optionals!

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,74 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+		return;
+
+	static std::regex const lineRE = std::regex(
+		R"~~~((^|\s+)@src\s+(\d+):(-1|\d+):(-1|\d+)\s*)~~~",
```
```suggestion
		R"~~~((^|\s+)@src\s+(\d+):(-1|\d+):(-1|\d+))(\s+|$)~~~",
```

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,74 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	auto const commentLocation = m_scanner->currentCommentLocation();
+	string const& text = m_scanner->currentCommentLiteral();
+	std::cmatch cm;
+	if (!std::regex_search(text.data(), text.data() + text.size(), cm, lineRE))
```
```suggestion
	if (!std::regex_search(text, cm, lineRE))
```

chriseth replied:
actually no need to define a local variable

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,74 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	auto const start = toInt(cm[3].str());
+	auto const end = toInt(cm[4].str());
+
+	if (!sourceIndex || !start || !end)
+	{
+		m_errorReporter.syntaxError(6367_error, commentLocation, "Invalid value in source location mapping. Could not parse location specification.");
+		m_locationOverride = SourceLocation{};
+		return;
+	}
+
+	if (!(*start <= *end))
+	{
+		m_errorReporter.syntaxError(5798_error, commentLocation, "Invalid value in source location mapping. Start offset larger than end offset.");
+		m_locationOverride = SourceLocation{};
+		return;
+	}
+
+	auto charStreamIt = m_charStreamMap->find(*sourceIndex);
+	if (charStreamIt == m_charStreamMap->end())
+	{
+		m_errorReporter.syntaxError(2674_error, commentLocation, "Invalid source mapping. Source index not defined via @use-src.");
+		m_locationOverride = SourceLocation{};
+		return;
+	}
+
+	shared_ptr<CharStream> charStream = charStreamIt->second;
+	solAssert(charStream, "");
+
+	m_locationOverride = SourceLocation{*start, *end, charStream};
```
```suggestion
	m_locationOverride = SourceLocation{};
	if (!sourceIndex || !start || !end)
		m_errorReporter.syntaxError(6367_error, commentLocation, "Invalid value in source location mapping. Could not parse location specification.");
	else if (!(*start <= *end))
		m_errorReporter.syntaxError(5798_error, commentLocation, "Invalid value in source location mapping. Start offset larger than end offset.");
	else if (!m_charStreamMap->count(*sourceIndex))
		m_errorReporter.syntaxError(2674_error, commentLocation, "Invalid source mapping. Source index not defined via @use-src.");
	else
	{
		shared_ptr<CharStream> charStream = m_charStreamMap->at(*sourceIndex);
		solAssert(charStream, "");

		m_locationOverride = SourceLocation{*start, *end, charStream};
	}
```

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,14 +47,38 @@ class Parser: public langutil::ParserBase
 		ParserBase(_errorReporter),
 		m_dialect(_dialect),
-		m_locationOverride(std::move(_locationOverride))
+		m_locationOverride{_locationOverride ? *_locationOverride :langutil::SourceLocation{}},
```
```suggestion
		m_locationOverride{_locationOverride ? *_locationOverride : langutil::SourceLocation{}},
```

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,14 +47,38 @@ class Parser: public langutil::ParserBase
+		m_useSourceLocationFrom{
+			_locationOverride ?
+				UseSourceLocationFrom::LocationOverride :
+				UseSourceLocationFrom::Scanner
```
```suggestion
			UseSourceLocationFrom::LocationOverride :
			UseSourceLocationFrom::Scanner
```

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -367,9 +452,10 @@ VariableDeclaration Parser::parseVariableDeclaration()
+			varDecl.debugData = updateLocationEndFrom(varDecl.debugData, locationOf(*varDecl.value));
 	}
-	else
+	if (m_useSourceLocationFrom == UseSourceLocationFrom::Scanner)
```
why not
```suggestion
	else if (m_useSourceLocationFrom == UseSourceLocationFrom::Scanner)
```
?

chriseth commented on libyul/AsmParser.h:
```diff
@@ -45,14 +47,38 @@ class Parser: public langutil::ParserBase
+		Scanner, LocationOverride, Comments,
+	};
+
+	/// Constructs a Yul parser.
```
```suggestion
```
completely useless information

chriseth commented on libyul/ObjectParser.cpp:
```diff
@@ -106,6 +106,7 @@ shared_ptr<Block> ObjectParser::parseCode()
 
 shared_ptr<Block> ObjectParser::parseBlock()
 {
+	// TODO: maybe here check for @use-src?
```
remove

chriseth commented:
Please add some tests where there source location changes in between statements like
```
{
  let x := /// @src ...
  add(1, /// @src ...
2)
}
```
and
```
{
  {
    sstore(0, 1)
    /// @src ...
  }
  mstore(1, 2)
}
```
(in the second example, only the source location of the mstore should be changed, not the block itself.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,62 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+		return;
+
+	static std::regex const lineRE = std::regex(
+		R"~~~((^|\s+)@src\s+(\d+):(-1|\d+):(-1|\d+)\s*$)~~~",
```
I think it's fine to use
```suggestion
		R"~~~((^|\s+)@src\s+(\d+):(-1|\d+):(-1|\d+)(\s+|$))~~~",
```

christianparpart replied:
This is caussing all tests to fail ;-(

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,62 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	);
+
+	auto const commentLocation = m_scanner->currentCommentLocation();
+	string const& text = m_scanner->currentCommentLiteral();
```
no need for a local variable

chriseth commented on libyul/AsmParser.h:
```diff
@@ -23,13 +23,15 @@
 #include <liblangutil/Scanner.h>
 #include <liblangutil/ParserBase.h>
 
+#include <functional>
```
```suggestion
```

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,62 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+
+	static std::regex const lineRE = std::regex(
+		//R"~~~((^|\s+)@src\s+(\d+):(-1|\d+):(-1|\d+)(\s+|$))~~~",
+		R"~~~((^|\s+)@src\s+(\d+):(-1|\d+):(-1|\d+)\s*\n*)~~~",
```
`\s*\n*` is useless

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,61 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	);
+
+	std::smatch matchResult;
+	if (!std::regex_search(m_scanner->currentCommentLiteral(), matchResult, lineRE))
```
We have to find the last match. Can you add a test for
```
/// @src 1:2:3
/// @src 4:5:7
sstore(0, 1)
```
The expected source location should be `4:5:7`

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,61 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	m_locationOverride = SourceLocation{};
+	if (!sourceIndex || !start || !end)
+		m_errorReporter.syntaxError(6367_error, commentLocation, "Invalid value in source location mapping. Could not parse location specification.");
+	else if (!(*start <= *end))
```
`-1` is valid

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +108,61 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+		m_errorReporter.syntaxError(6367_error, commentLocation, "Invalid value in source location mapping. Could not parse location specification.");
+	else if (!(*start <= *end))
+		m_errorReporter.syntaxError(5798_error, commentLocation, "Invalid value in source location mapping. Start offset larger than end offset.");
+	else if (!m_charStreamMap->count(*sourceIndex))
```
`-1` should be valid here as well.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +93,67 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+	);
+
+	string text = m_scanner->currentCommentLiteral();
+	while (!text.empty())
```
What about using `std::sregex_iterator`?

christianparpart replied:
Oh I didn't know about that!

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -78,14 +93,67 @@ unique_ptr<Block> Parser::parse(std::shared_ptr<Scanner> const& _scanner, bool _
+			m_errorReporter.syntaxError(5798_error, commentLocation, "Invalid value in source location mapping. Start offset larger than end offset.");
+		else if (!(sourceIndex >= 0 && m_charStreamMap->count(static_cast<unsigned>(*sourceIndex))))
+			m_errorReporter.syntaxError(2674_error, commentLocation, "Invalid source mapping. Source index not defined via @use-src.");
+		else if (sourceIndex >= 0)
```
I'm wondering if we should still allow `-1:0:10` to lead to a "real" source location (instead of the default-constructed one)

christianparpart replied:
hm, I think that sounds like an interesting thing. I can do that. It would again require some more test cases, I think.

chriseth replied:
Then please do.

chriseth commented on libyul/AsmParser.h:
```diff
@@ -96,7 +126,10 @@ class Parser: public langutil::ParserBase
-	std::optional<langutil::SourceLocation> m_locationOverride;
+
+	std::optional<std::map<unsigned, std::shared_ptr<langutil::CharStream>>> m_charStreamMap;
+	langutil::SourceLocation m_locationOverride;
```
Can you change this to a `shared_ptr`? If it stays like this, we will create tons of different shared pointers pointing at equivalent data.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -107,6 +175,8 @@ Statement Parser::parseStatement()
 		_if.condition = make_unique<Expression>(parseExpression());
 		_if.body = parseBlock();
+		if (m_useSourceLocationFrom == UseSourceLocationFrom::Scanner)
+			_if.debugData = updateLocationEndFrom(_if.debugData, currentLocation());
```
```suggestion
			_if.debugData = updateLocationEndFrom(_if.debugData, _if.body.debugData);
```

chriseth replied:
The current location already points to the next token after the block.

christianparpart committed:
Adapted tests due to more precise Yul source locations.

Also added support for -1 source index, referencing original scanner's source location.

christianparpart commented:
@chriseth there are two other note-worthy / important changes.

1. the source location of an `if` statement now covers all until the end of the body (initially only the `if` token), and
2. the source location of a block now also covers the full block.

these two changes are the reason for the test case changes that are not in `test/yul/Parser.cpp`.

christianparpart committed:
Use shared DebugData for when using source locations from comments.

chriseth commented on libyul/AsmParser.cpp:
```diff
@@ -128,21 +128,22 @@ void Parser::fetchSourceLocationFromComment()
 			m_errorReporter.syntaxError(5798_error, commentLocation, "Invalid value in source location mapping. Start offset larger than end offset.");
 		else if (sourceIndex == -1 && (0 <= *start && *start <= *end)) // Use source index -1 to indicate original source.
-			m_locationOverride = SourceLocation{*start, *end, ParserBase::currentLocation().source};
+		{
+			m_debugDataOverride = DebugData::create(SourceLocation{*start, *end, ParserBase::currentLocation().source});
+		}
```
```suggestion
			m_debugDataOverride = DebugData::create(SourceLocation{*start, *end, ParserBase::currentLocation().source});
```

chriseth replied:
Actually I think it would be better to keep a single `DebugData` struct and only update its `sourceLocation` member when we encounter `@src`. This will make it easier to add more tags in the future, since it does not reset the other members all the time.
